### ISSUE TITLE
External diff command do not support `columnWidth` placeholder

### ISSUE THREAD

**Issue by AFutureD** [jesseduffield/lazygit#3119]
**Describe the bug**

Using `difftastic` in external diff command with `side-by-side` display type, there is a view problem.
The lines in the right view unexpectedly wrapped.

**To Reproduce**
Simply scustom external diff config.

```
git:
  paging:
    externalDiffCommand: difft --color=always --display=side-by-side --syntax-highlight=off
```

**Expected behavior**
Pass window's width to external diff command like `pager`.

**Screenshots**
![CleanShot 2023-11-19 at 13 57 05@2x](https://github.com/jesseduffield/lazygit/assets/17228538/594a7cc4-1630-4139-9a2d-53402c9fcab3)

**Version info:**
```
> lazygit --version
commit=1d1b8cc01f87bb3495426ac8d81d97573f6840d4, build date=2023-10-27T04:14:57Z, build source=unknown, version=1d1b8cc0, os=darwin, arch=arm64, git version=2.39.2 (Apple Git-143)
> git --version
git version 2.39.2 (Apple Git-143)
```
**Additional context**

**Note:** please try updating to the latest version or [manually building](https://github.com/jesseduffield/lazygit/#manual) the latest `master` to see if the issue still occurs.

**Comment by stefanhaller** [comment]
Lazygit is already passing the window width to the external diff command by means of the COLUMNS environment variable, so most of the time the layout should be correct. There is a bug, however, when lazygit's view configuration changes, for example when you switch from a file that has only unstaged changes (one wide main view) to a file that has both staged and unstaged changes (main view is split into two more narrow ones). In this situation, lazygit will call the diff command with the _old_ window width one time. If you then hit `<shift-R>` to refresh (or wait 10 seconds for an auto refresh), the display should be correct again.

You can confirm this behavior by setting your external diff command to `externalDiffCommand: /bin/sh -c "echo $COLUMNS"` for testing.

I'm not sure how easy it will be to change the code so that diff commands are always called with the up-to-date window dimensions after the view layout changed. I'll have a look.

**Comment by AFutureD** [comment]
Hi, I have tried set external diff command to externalDiffCommand: /bin/sh -c "echo $COLUMNS".
But, there is nothing shows on screen, and I have changed to echo other String which show well.
![CleanShot 2023-11-20 at 09 32 50@2x](https://github.com/jesseduffield/lazygit/assets/17228538/5e860deb-555d-448c-8ac0-f873d7fd99e2)


And I have also tried to wait (more than 10s) screen refresh by itself, there is nothing change.
By setting externalDiffCommand to `date`, I'm sure screen did refresh as expected, but COLUMNS was empty.

**Comment by stefanhaller** [comment]
Wow, interesting. That's a different problem then; we'll have to find out why COLUMNS isn't set on your system.

Anybody who is more experienced with terminal stuff and PTYs have an idea what could be wrong here? I don't know much about these topics.

I'd be curious: what does `tput columns` say in your shell, and what do you get in lazygit if you set `externalDiffCommand: tput columns`?

**Comment by AFutureD** [comment]
I'm using macOS, so I set the externalDiffCommand to `tput cols`.
The result is `80`, and now change when I resize terminal window or do other things like hit `<shift-R>`.

**Comment by AFutureD** [comment]
I have some time yesterday, so I tried to debug lazygit.
Here is what I found:
1. I do not have any pager settings.
2. Command depend on `gui.desiredPtySize` to refresh some info include view size.
3. Command without pager will than use `gui.newCmdTask` instead of `gui.newPtyTask`

So, when `ShowCmdObj` tries to run, all controllers using `gui.newPtyTask` to execute.
But when the task invoke,   the [if condition](https://github.com/jesseduffield/lazygit/blob/bb87642aee6bf6f599607e771d7a238a21eb4d81/pkg/gui/pty.go#L49C10-L49C10), find there is no page , so it using  `gui.newCmdTask` instead, and view info now pass to the environment.

I try [this change](https://github.com/jesseduffield/lazygit/pull/3120/commits/31ae94f3c190dfe3132db261be292427c34abbb8) it works well.

**Comment by AFutureD** [comment]
https://github.com/jesseduffield/lazygit/pull/3120/commits/11b87a69f83bbabe54fd273ff3ed5fedef7d7cf1

Question: Why lazygit set `--no-ext-diff` by default ?

For now, difft can use be used directly without set `--ext-diff`. See: https://difftastic.wilfred.me.uk/git.html#regular-usage

**Comment by AFutureD** [comment]
PR #3120 is ready.

But I wonder if such problem can write some unit test, so that if will not happened again.

**Comment by stefanhaller** [comment]
Thanks. This PR is now very confusing, because the PR title and description no longer matches the change you are making. I thought it would have been easier to open a new PR instead, but I'd also be ok with changing the title and description to match.

**Comment by stefanhaller** [comment]
As for testing: I can't think of a good way to test this in a unit test, so I'd be ok with not adding one.

**Comment by stefanhaller** [comment]
Fixed by #3120.
